cli/flox-rust-sdk/src/providers/buildenv.rs

```diff
@@ -94,77 +119,373 @@ pub trait BuildEnv {
+    /// this function is currently assumes that the package is from the nixpkgs base-catalog.
+    /// Currently the type is distinguished by the [LockedPackageCatalog::locked_url].
+    /// If this does not indicate a nixpkgs package, the function will currently panic!
+    fn realise_nixpkgs(&self, locked: &LockedPackageCatalog) -> Result<(), BuildEnvError> {
+        // check if all store paths are valid, or can be substituted
+        // if so, return without eval
+        let all_valid = self.check_store_path_with_substituters(locked.outputs.values())?;
+
+        if all_valid {
+            return Ok(());
+        }
+
+        let mut nix_build_command = self.base_command();
 
-        debug!(cmd=%pkgdb_realise_cmd.display(), "realising packages");
-        match call_pkgdb(pkgdb_realise_cmd, false) {
-            Ok(_) => {},
-            Err(CallPkgDbError::PkgDbError(err)) => return Err(BuildEnvError::Realise(err)),
-            Err(err) => return Err(BuildEnvError::CallPkgDb(err)),
+        // for now assume the plugin is located relative to the pkgdb binary
+        // <pkgdb>
+        // ├── bin
+        // │   └── pkgdb
+        // └── lib
+        //     └── nix-plugins
+        //          └── wrapped-nixpkgs-input.(so|dylib)
+        {
+            let pkgdb_lib_dir = Path::new(&*PKGDB_BIN)
+                .ancestors()
+                .nth(2)
+                .expect("pkgdb is in '<store-path>/bin'")
+                .join("lib/nix-plugins");
+
+            nix_build_command.args([
+                "--option",
+                "extra-plugin-files",
+                &pkgdb_lib_dir.to_string_lossy(),
+            ]);
+        }
+
+        let installable = {
+            let mut locked_url = locked.locked_url.to_string();
+
+            if let Some(revision_suffix) = locked_url.strip_prefix(NIXPKGS_CATALOG_URL_PREFIX) {
+                locked_url = format!("{FLOX_NIXPKGS_PROXY_FLAKE_REF_BASE}/{revision_suffix}");
+            } else {
+                todo!(
+                    "Building non-nixpkgs catalog packages is not yet supported.\n\
+                    Pending implementation and decisions regarding representation in the lockfile"
+                );
+            }
+
+            // build all out paths
+            let attrpath = format!("legacyPackages.{}.{}^*", locked.system, locked.attr_path);
+
+            format!("{}#{}", locked_url, attrpath)
+        };
+
+        nix_build_command.arg("build");
+        nix_build_command.arg("--no-write-lock-file");
+        nix_build_command.arg("--no-update-lock-file");
+        nix_build_command.args(["--option", "pure-eval", "true"]);
+        nix_build_command.arg("--no-link");
+        nix_build_command.arg(&installable);
+
+        debug!(%installable, cmd=%nix_build_command.display(), "building catalog package:");
+
+        let output = nix_build_command
+            .output()
+            .map_err(BuildEnvError::CallNixBuild)?;
+
+        if !output.status.success() {
```

Comment:
followup question: does this regress any of our errors, e.g. PACKAGE_BUILD_FAILURE? If not, is all the error handling code dead code and we can drop it?